plugin/path_checkouts.go

```diff
@@ -5,11 +5,349 @@ import (
+	}
+}
+
+func (b *backend) operationSetManageCheckIn(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	return b.handleCheckIn(ctx, req, fieldData, true)
+}
+
+func (b *backend) handleCheckIn(ctx context.Context, req *logical.Request, fieldData *framework.FieldData, overrideCheckInEnforcement bool) (*logical.Response, error) {
+	setName := fieldData.Get("name").(string)
+
+	serviceAccountNamesRaw, serviceAccountNamesSent := fieldData.GetOk("service_account_names")
+	var serviceAccountNames []string
+	if serviceAccountNamesSent {
+		serviceAccountNames = serviceAccountNamesRaw.([]string)
+	}
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+	if set == nil {
+		return logical.ErrorResponse(`"%s" doesn't exist`, setName), nil
+	}
+
+	// If check-in enforcement is overridden or disabled at the set level, we should consider it disabled.
+	disableCheckInEnforcement := overrideCheckInEnforcement || set.DisableCheckInEnforcement
+
+	// Track the service accounts we check in so we can include it in our response.
+	var checkIns []string
+
+	// Build and validate a list of service account names that we will be checking in.
+	var cantCheckInErrs error
+	if len(serviceAccountNames) == 0 {
+		// It's okay if the caller doesn't tell us which service accounts they
+		// want to check in as long as they only have one checked out.
+		// We'll assume that's the one they want to check in.
+		toCheckIn := make(map[string]*locksutil.LockEntry)
+		for _, setServiceAccount := range set.ServiceAccountNames {
+			// This is called in a func to avoid calling defer in a loop.
+			if err := func() error {
+				lock := locksutil.LockForKey(b.checkOutLocks, setServiceAccount)
+				lock.Lock()
+				releaseLock := true
+				defer func() {
+					if releaseLock {
+						lock.Unlock()
+					}
+				}()
+
+				checkOut, err := b.checkOutHandler.Status(ctx, req.Storage, setServiceAccount)
+				if err != nil {
+					return err
+				}
+				if checkOut.IsAvailable {
+					return nil
+				}
+				if !canCheckIn(req, checkOut, disableCheckInEnforcement) {
+					return nil
+				}
+				releaseLock = false
+				toCheckIn[setServiceAccount] = lock
+				return nil
+			}(); err != nil {
+				return nil, err
+			}
+		}
+		defer func() {
+			for _, lock := range toCheckIn {
+				lock.Unlock()
+			}
+		}()
+		switch len(toCheckIn) {
+		case 0:
+			// Everything's already currently checked in, nothing else to do here.
+		case 1:
+			// This is what we need to continue with the check-in process.
+			for serviceAccountName, _ := range toCheckIn {
+				if err := b.checkOutHandler.CheckIn(ctx, req.Storage, serviceAccountName); err != nil {
+					return nil, err
+				}
+				checkIns = append(checkIns, serviceAccountName)
+			}
+		default:
+			return logical.ErrorResponse(`when multiple service accounts are checked out, the "service_account_names" to check in must be provided`), nil
+		}
+	} else {
```

Comment:
I think we may be able to refactor out the first half of this `if/else`. Since the second half knows how to handle checking in given a list of serviceAccountNames, can we refactor the `len(serviceAccountNames) == 0` block to just generate a list of names to feed to the second part? In other words, in the user doesn't provide the list of names, check to see if it is just one name in the set, and if so use that as the list of names.

Reply:
I will take a whack at that. Not sure how it'll turn out with the additional changes of the set-level locking, we'll see.

Reply:
I tried it out and it gets really weird with a lot of interwoven if/else logic. I'd much prefer to leave it as-is because to me it seems cognitively clearer.

Reply:
yeah, it sort of does now that I look at it closer. Too bad, as there is a lot of duplication but I can't think of a cleaner refactor.